### ISSUE TITLE
redirect multiprocess hangs sometimes for large batch file on windows 2012 server

### ISSUE THREAD

**Issue by giko45** [taers232c/GAMADV-XTD3#29]
Hi,

I am executing the following command
```gam.exe redirect stdout stdout.txt multiprocess redirect stderr stderr.txt multiprocess batch test.csv```

where test.csv contains about 31000 lines of
```
gam print group-members group grp_email_1@domain.com
gam print group-members group grp_email_2@domain.com
...
gam print group-members group grp_email_30714@domain.com
```

This takes normally about 30 minutes (with 10 threads).

Sometimes (about 1 out of 10-20) the command hangs with the following symtoms
* stderr.txt and stdout.txt are created  with 0 bytes.
* process manager shows all gam.exe processes
* when i kill the gam.exe process tree (all gam processes are killed) both stderr.txt and stdout.txt are written and filled with normal output (so all group members are listed in the stdout.txt file and all "Getting all Members, Managers, Owners for..." lines are listed in the tsderr.txt file ) except for the last lines in stderr.txt 
```
Using 10 processes...
Processing item 100
....
Processing item 30700

```
which are missing.

So it seems that the "master" gam process hangs somewere after all subprocesses are finnished normally and theire stdout+stderr are writen to the output/stderr streams  (they show 0 bytes but are locked and probably alread written)  but before writing its own stderr to the output stream. 

How can i circumvent this error or debug this?

I am running this on a windows 2012 server

**Comment by taers232c** [comment]
I'll have to take a look at this
Gam uses a pool of subprocesses; when one finishes, gam assigns the next task to it. When all tasks have been assigned, gam waits for all process to finish and then exits. If one child process hung you'd expect to see one parent and one child, not a bunch of children. Let me see what I can ad in the way of additional debugging.

**Comment by giko45** [comment]
Otherwise i can just chunk the input file to see if that makes it more stable.

**Comment by taers232c** [comment]
Call me on Google Hangouts or send a Meet invitation.

ross.scroggs@gmail.com

**Comment by taers232c** [comment]
This won't solve the problem but is a simplification and considerably reduces the size of test.csv
```
$ more test.csv
group
testgroup1@domain.com
testgroup2@domain.com
$ gam redirect stdout stdout.txt multiprocess redirect stderr stderr.txt multiprocess csv text.csv gam print group-members group ~group
Using 2 processes...
```

**Comment by giko45** [comment]
I will change the csv file per your suggestion. Since the error occures intermittent it might take a while before i can see a change.

**Comment by taers232c** [comment]
Changing to the CSV form only simplifies your data, behind the scenes it is the same processing, so it is not going to solve the problem. I don't yet see where the hang comes from.

**Comment by giko45** [comment]
ok, I will also split up the file in smaler parts to see if i can connect the hang to a specific group.

**Comment by giko45** [comment]
just a wild guess:  https://bugs.python.org/issue37424  scroll down to "https://bugs.python.org/issue37424"

**Comment by taers232c** [comment]
I don't think that's it; anyway if you have a current version GAM is using Python 3.8.1.
Is there any change you can temporarily install GAM on Linux/MacOS so we can see if this is a Windows issue.

**Comment by taers232c** [comment]
This should be fixed in 5.08+
